cli: deprecate `jj checkout` and `jj merge`

thoughtpolice opened the pull request:
As discussed on Discord and in previously in #1713, this deprecates `jj checkout` and `jj merge`, as they are simply specializations for existing functionality under `jj new`.

Both commands are now hidden by default from `help`.

Supersedes #1713.

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [x] I have added tests to cover my changes

martinvonz commented on cli/tests/test_git_colocated.rs:
```diff
@@ -130,6 +130,8 @@ fn test_git_colocated_unborn_branch() {
     let (stdout, stderr) = test_env.jj_cmd_ok(&workspace_root, &["checkout", "root()"]);
```
nit: Update this test case to use `jj new` instead? it's not meant to test `jj checkout`.

The same is true for part of the `test_checkout_conflicting_branches()` and `test_checkout_conflicting_change_ids()` above. I think we'd ideally move those to `test_conflict.rs` before this commit, but I won't insist if you don't have time for it.

martinvonz replied:
Oh, sorry to have wasted your time, moving them to `test_new_command.rs` is what I meant!

martinvonz commented on CHANGELOG.md:
```diff
@@ -28,6 +28,26 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   **Deadline**: `jj checkout` will be deleted and become a **hard error in
   version 0.18.0**.
 
+* `jj merge` is now deprecated; use `jj new` instead.
+
+  **Rationale**: `jj merge` is functionally equivalent to `jj new`, except it
+  requires *at least* two arguments. `jj new` can create a new commit with any
+  number of parents; whether that is 1 parent &mdash; a single commit &mdash; or
+  2, or 3 parents, which would represent various merges. `jj merge` is a
+  specialization of existing functionality; it only creates "merge commits", but
+  only because "merge commit" is defined as "has at least two parents."
+
+  This makes `jj merge` functionally redundant, and also obscures the relative
+  generality of the `jj new` command, which can create a commit from any number
+  of parents. We encourage all users to adopt `jj new` for performing merges
+  instead; it is shorter, more general, and easier to remember.
+
+  `jj merge` will no longer be shown as part of `jj help`, but will still
+  function for now, emitting a warning about its deprecation.
+
+  **Deadline**: `jj merge` will be deleted and become a **hard error in
+  version 0.18.0**.
```
It seems simpler to merge this with the item about `jj checkout` above, if you think we can make it readable enough (seems like it to me). Then we don't waste the reader's time as much by making them read it twice (or, perhaps more likely, making them compare the two sections to see if there's any surprising difference between them).

martinvonz commented:
> I chose jj 0.18.0 as the version to target for removal, as the next release is 0.14, and we do monthly releases, resulting in a four-month window. I think this is a fair amount of time, but I'm open to other suggestions if people want it to be longer or shorter.

I think we've aimed for at least 6 months before. I see little harm in giving users a bit longer. It's cheap to maintain, and users are probably not deliberately going to keep using the old commands if we warn every time as we do after this PR. What do you think?

PhilipMetzger commented on cli/src/commands/checkout.rs:
```diff
@@ -41,6 +41,14 @@ pub(crate) fn cmd_checkout(
+        ui.warning(),
+        "warning: `jj checkout` is deprecated; use `jj new` instead, which is equivalent"
+    )?;
+    writeln!(
+        ui.warning(),
+        "warning: `jj checkout` will be removed in jj 0.18, and this will be a hard error"
+    )?;
```
I'd remove this as this gives users something to hold us accountable for.

PhilipMetzger commented on CHANGELOG.md:
```diff
@@ -7,6 +7,27 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+
+* `jj checkout` is now deprecated; use `jj new` instead.
+
+  **Rationale**: `jj checkout` is functionally equivalent to `jj new`, except it
+  can only take a single argument; `jj new` can take more than one argument,
+  resulting in a merge commit with multiple parents. They are always equivalent
+  in the case of a single argument, hence `jj new` is a strict superset of
+  functionality.
+  
+  More importantly, the verb `checkout` is a relic of a bygone era &mdash; the
+  days of RCS file locking, before 3-way merge &mdash; and is not a good,
+  fitting name for the functionality it provides. We encourage all users to
+  adopt `jj new` instead; it is shorter, more general, and easier to remember.
+
+  `jj checkout` will no longer be shown as part of `jj help`, but will still
+  function for now, emitting a warning about its deprecation.
+
+  **Deadline**: `jj checkout` will be deleted and become a **hard error in
+  version 0.18.0**.
```
Nit: Remove the deadline and move a part of the rationale to the commit message. I'd say the explanation of the checkout verb should definitely be in the commit message.

thoughtpolice replied:
Not sure how to word this in such a short space or if it's needed; will think on it...

PhilipMetzger replied:
> Not sure how to word this in such a short space

Who said that that the commit message should be short? I'd like to see an TLDR and then you can an even more extensible explananation behind for all I care.

thoughtpolice replied:
Oh, for some reason I thought you said *warning* message, not commit message.

PhilipMetzger commented on cli/src/commands/merge.rs:
```diff
@@ -24,6 +24,14 @@ pub(crate) fn cmd_merge(
+        ui.warning(),
+        "warning: `jj merge` is deprecated; use `jj new` instead, which is equivalent"
+    )?;
+    writeln!(
+        ui.warning(),
+        "warning: `jj merge` will be removed in jj 0.18, and this will be a hard error"
+    )?;
```
Ditto

PhilipMetzger reviewed (commented):
Thanks for finally doing finishing this :-). I've left some comments. A final question is what do we do with `jj checkout`'s aliases, do they stay as convience aliases or are they just removed? 
This was something we didn't decide on the last time the deprecation was brought up in Discord, IIRC ilya is a big user of `jj co`.

thoughtpolice commented:
I think 6 months is fine, so I adjusted it the timeline in the CHANGELOG.

But I removed the exact notice of version in the error messages as Philip asked; it's only now mentioned in the changelog. I feel this is a good balance, since it's not something people will precisely beat us over the head with I think, and the actual command itself isn't too committal.

I'm not sure about adding a rationale to the actual message; I don't think making it too long helps. I'll think about it a tiny bit more.

thoughtpolice commented:
> A final question is what do we do with `jj checkout`'s aliases, do they stay as convience aliases or are they just removed? This was something we didn't decide on the last time the deprecation was brought up in Discord, IIRC ilya is a big user of `jj co`.

We could remove them when the commands are actually deleted, so that `jj checkout` is an error but `jj co` is free &mdash; which means Ilya would be able to re-purpose `jj co` again as an alias. That might be an OK tradeoff.

Right now, aliases cannot override any command even if it's a no-op. So we'd otherwise have to fix that.

martinvonz commented on cli/tests/test_git_colocated.rs:
```diff
@@ -191,6 +191,8 @@ fn test_git_colocated_unborn_branch() {
     let (stdout, stderr) = test_env.jj_cmd_ok(&workspace_root, &["checkout", "root()"]);
```
nit: use `new` here too

martinvonz reviewed (approved):
Thanks!

PhilipMetzger commented:
> We could remove them when the commands are actually turned into hard errors, allowing them to be reused; so that `jj checkout` is an error but `jj co` is free — which means Ilya would be able to re-purpose `jj co` again as an alias. That might be an OK tradeoff.
> 
> Right now, aliases cannot override any command even if it's a no-op. So we'd otherwise have to fix that.

Basically https://github.com/martinvonz/jj/pull/1713#issuecomment-1597785456 and the following two messages seem a more appropriate approach. I think ATM moving the aliases to `jj new` makes sense and after actually removing `jj checkout` we can discuss the removal of the aliases. 

My argument for the whole thing is that a CLI is basically a API, so the same concepts should apply. 

And another thing which I only noticed now, why is it a warning instead of an hard error like #1911?

Fraetor commented on CHANGELOG.md:
```diff
@@ -7,6 +7,43 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+  `jj checkout` and `jj merge` will no longer be shown as part of `jj help`, but
+  will still function for now, emitting a warning about its deprecation.
+
+  **Deadline**: `jj checkout` will be deleted and is expected become a **hard
```
Should `jj merge` also have a deadline here?

Fraetor commented:
I've spotted two places in the documentation where checkout is mentioned.

- `jj co main` on the [Git comparison page](https://martinvonz.github.io/jj/prerelease/git-comparison/).
- `jj checkout/co` in the [Tutorial](https://martinvonz.github.io/jj/prerelease/tutorial/#creating-our-first-change).

ilyagr commented on CHANGELOG.md:
```diff
@@ -7,6 +7,43 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+  or even two commands to remember for all these cases &mdash; it is much
+  simpler to have just one for every case.
+
+  We encourage all users to adopt `jj new` instead; it is shorter, more general,
```
Minor nit: `jj co` is shorter than `jj new`. You can just get rid of the "it is shorter" clause.

My "you could alias..." suggestion above could also go here, if you don't replace the entire Rationale with it as I suggested above.

ilyagr commented on CHANGELOG.md:
```diff
@@ -7,6 +7,43 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+* `jj checkout` and `jj merge` are both deprecated; use `jj new` instead to
+  replace both of these commands in all instances.
+
+  **Rationale**: `jj checkout` and `jj merge` implement two distinct pieces of
```
Optional: I feel like this is not a great place to have a detailed rationale. Perhaps you could just put it into the PR description, and say "See <link> for rationale"? I don't insist on this.

ilyagr replied:
You could instead say 

> you can alias "co", "checkout", and/or "merge" to "new" if you prefer the old behavior.

ilyagr replied:
I just realized this is already discussed in https://github.com/martinvonz/jj/pull/2842#discussion_r1457848112.

ilyagr reviewed (commented):
Left some very minor and optional nits about the changelog.

I think this is a good change, hopefully the users will agree.

thoughtpolice commented:
> This seems very close to done. Do you think you can find time to get this merged before I cut a new release next week?

Yeah, I should be able to do it tonight or this weekend I think.

thoughtpolice commented:
I think I've addressed everyone's nits; however, I have *still* left the rationale in the changelog, and the reason is simple: because users read the changelog, but we read the history.

I have however made the actual changelog entry much shorter and more direct, and I've also put the more detailed rational in the commit messages themselves as requested. So I think this is a good end result all around.

martinvonz reviewed (approved):
Thanks!

martinvonz commented:
I sent https://github.com/martinvonz/jj/pull/2929 to unblock CI

ilyagr commented:
(Non-blocking)

> We could remove them when the commands are actually turned into hard errors, allowing them to be reused; so that jj checkout is an error but jj co is free — which means Ilya would be able to re-purpose jj co again as an alias. That might be an OK tradeoff.

I think I'd literally make them into default aliases in the built-in settings. Then, people can change them. But this is not important and a discussion for another time.

thoughtpolice commented:
That's a good idea; we can still follow up on that in time for the release window, I think. But yes, I'll go ahead and merge this.

matts1 commented:
Was chatting with @martinvonz today about the warning we now show.

```
warning: `jj checkout` is deprecated; use `jj new` instead, which is equivalent
warning: `jj checkout` will be removed in a future version, and this will be a hard error
```

 `jj checkout`, `jj amend`, `jj merge`, and `jj commit`, while not actually required, all allow new jj users to think with their git-brain rather than having to come up with a whole new understanding of how `jj` works. I'm all for showing a message whenever you use these commands suggesting how to use things in what I'll call "jj-style", but personally I strongly advocate against removing `jj checkout` and similar commands, since I believe doing so will increase the barrier-to-entry of jj with no explicit benefits.

> I think I'd literally make them into default aliases in the built-in settings. Then, people can change them.

This seems like a reasonable approach to me. That being said, it did confuse an internal user when they wrote `jj help amend` and got documentation for squash. I think the approach I'd like is for them to be a default alias, and as a seperate feature request, when I run `jj help <alias>`, it would expand the alias first. Something like:

```
jj help checkout
Alias for jj new

Help for jj new:
<jj help new>
```

ilyagr commented:
> That being said, it did confuse an internal user when they wrote jj help amend and got documentation for squash. I think the approach I'd like is for them to be a default alias, and as a seperate feature request, when I run jj help <alias>, it would expand the alias first. 

I did request the same thing in https://github.com/martinvonz/jj/issues/2866#issuecomment-1904777613. I'm guessing @jyn514 might be working on a simpler version of that feature first, though.

PhilipMetzger commented:
> Was chatting with @martinvonz today about the warning we now show.
> 
> ```
> warning: `jj checkout` is deprecated; use `jj new` instead, which is equivalent
> warning: `jj checkout` will be removed in a future version, and this will be a hard error
> ```
> 
> `jj checkout`, `jj amend`, `jj merge`, and `jj commit`, while not actually required, all allow new jj users to think with their git-brain rather than having to come up with a whole new understanding of how `jj` works. I'm all for showing a message whenever you use these commands suggesting how to use things in what I'll call "jj-style", but personally I strongly advocate against removing `jj checkout` and similar commands, since I believe doing so will increase the barrier-to-entry of jj with no explicit benefits.

I am strongly against upholding the whole git shtick, as removing these commands makes `jj` unique and allows us to have some affordances on behalf of the users workflow. While Git is the dominant system for the forseeable future, it shouldn't prevent innovation and UI affordances like `jj` already has. Or do you also want to go back to manually `jj add`ing files? 

I also dislike that this concern was raised so late after we discussed the removal for so long.

martinvonz commented:
> While Git is the dominant system for the forseeable future, it shouldn't prevent innovation and UI affordances like `jj` already has.

I don't think keeping `jj checkout` and `jj merge` as almost-aliases for `jj new` really prevents any innovation, unless we feel like we want to use those command names for some new feature. The commands print a warning and point the user to `jj new`, so I think we're making it clear that we think there's a better way.

> Or do you also want to go back to manually `jj add`ing files?

That's #323. Let's keep this discussion focused on `jj new/checkout/merge`.

> I also dislike that this concern was raised so late after we discussed the removal for so long.

Not all users are going to be part of our Discord and/or GitHub discussions, so they will not notice until they upgrade to a release with the new behavior.

joyously commented:
To be complete: I don't think to use `checkout`, but hearing of it makes me think of SVN, even though I use Breezy with "lightweight checkout" branches.
As for `merge`, in Breezy it's for branches(which are the main focus), and there aren't really branches in `jj` (similar to Darcs) so I don't think of it.

PhilipMetzger commented:
> I don't think keeping `jj checkout` and `jj merge` as almost-aliases for `jj new` really prevents any innovation, unless we feel like we want to use those command names for some new feature. The commands print a warning and point the user to `jj new`, so I think we're making it clear that we think there's a better way.

My main point is being _unique_ with _not_ having these commands at all. The best moment to remove something is yesterday, today or not at all. 

> > Or do you also want to go back to manually `jj add`ing files?
> 
> That's #323. Let's keep this discussion focused on `jj new/checkout/merge`.

This was a purely rhetorical question, in terms of going back to the _old ways_ because innovation shouldn't happen. 

> > I also dislike that this concern was raised so late after we discussed the removal for so long.
> 
> Not all users are going to be part of our Discord and/or GitHub discussions, so they will not notice until they upgrade to a release with the new behavior.

Matt is atleast in the Discord, not a dogfooded user which I never heard of, so there should atleast be a better way to present internal feedback to OSS users instead of just backing out something with a unreasonable short explanation.

martinvonz commented:
> there should atleast be a better way to present internal feedback to OSS users instead of just backing out something with a unreasonable short explanation and response time.

I tried to explain it in https://github.com/martinvonz/jj/pull/3218/commits/97d1cb756c26b58f290b471dedfc2dbd991e9de9. I'm sorry that it felt unreasonably short. What additional information would you have wanted me to include? Or maybe your objection is to using a PR for discussion? Oh, I should probably have included "RFD:" in the PR title? I'll update it.

joyously commented:
Both of them seemed fast to me, especially the second one.
Was it based on that one internal user?
I am one user, and to me, `checkout` means `svn checkout` which gives you just a piece (one folder) of the repo or `brz checkout` which gives you a branch (which in Breezy is a folder).
I don't like how several VCS have used the same term differently, and adding one more is not that great an idea.

martinvonz commented:
> Was it based on that one internal user?

No, it was based on feedback I've heard from several people over many months.

> I am one user, and to me, `checkout` means `svn checkout` which gives you just a piece (one folder) of the repo or `brz checkout` which gives you a branch (which in Breezy is a folder).
> I don't like how several VCS have used the same term differently, and adding one more is not that great an idea.

I'm sorry, but I think you're an unusual user :) Most users come from Git, and Git does have a `checkout` that's very very similar to `jj checkout`.

ilyagr commented:
> I did request the same thing in https://github.com/martinvonz/jj/issues/2866#issuecomment-1904777613. I'm guessing @jyn514 might be working on a simpler version of that feature first, though.

Actually, there's already https://github.com/martinvonz/jj/pull/3015 that can be looked at.
